agents/src/transcription.ts

```diff
@@ -1,134 +1,308 @@
+  pushAudio(frame: AudioFrame) {
+    this.#checkNotClosed();
+    if (!this.#audioData) {
+      this.#audioData = { pushedDuration: 0, done: false };
+      this.#audioQ.push(this.#audioData);
+      this.#audioQChanged.put(1);
+    }
+    this.#audioData.pushedDuration += frame.samplesPerChannel / frame.sampleRate;
   }
 
-  start(): void {
-    if (!this.#isRunning) {
-      this.#isRunning = true;
-      this.#startPublishingLoop().catch((error) => {
-        this.#logger.error('Error in publishing loop:', error);
-        this.#isRunning = false;
-      });
+  pushText(text: string) {
+    this.#checkNotClosed();
+    if (!this.#textData) {
+      this.#textData = {
+        sentenceStream: this.#opts.sentenceTokenizer.stream(),
+        pushedText: '',
+        done: false,
+        forwardedHyphens: 0,
+        forwardedSentences: 0,
+      };
+      this.#textQ.push(this.#textData);
+      this.#textQChanged.put(1);
     }
+
+    this.#textData.pushedText += text;
+    this.#textData.sentenceStream.pushText(text);
   }
 
-  pushAudio(frame: AudioFrame): void {
-    this.#totalAudioDuration += frame.samplesPerChannel / frame.sampleRate;
+  markAudioSegmentEnd() {
+    this.#checkNotClosed();
+
+    if (this.#audioData) {
+      // create empty audio data if none exists
+      this.pushAudio(new AudioFrame(new Int16Array(), 24000, 1, 0));
+    }
+
+    this.#audioData!.done = true;
+    this.#audioData = undefined;
   }
 
-  pushText(text: string): void {
-    this.#currentText += text;
+  markTextSegmentEnd() {
+    this.#checkNotClosed();
+
+    if (!this.#textData) {
+      this.pushText('');
+    }
+
+    this.#textData!.done = true;
+    this.#textData?.sentenceStream.flush();
+    this.#textData?.sentenceStream.close();
+    this.#textData = undefined;
   }
 
-  #textIsComplete: boolean = false;
-  #audioIsComplete: boolean = false;
+  segmentPlayoutStarted() {
+    this.#checkNotClosed();
+    this.#playingSegIndex++;
 
-  markTextComplete(): void {
-    this.#textIsComplete = true;
-    this.#adjustTimingIfBothFinished();
+    if (!this.#task) {
+      this.#task = this.#mainLoop();
+    }
   }
 
-  markAudioComplete(): void {
-    this.#audioIsComplete = true;
-    this.#adjustTimingIfBothFinished();
+  segmentPlayoutFinished() {
+    this.#checkNotClosed();
+    this.#finishedSegIndex++;
   }
 
-  #adjustTimingIfBothFinished(): void {
-    if (this.#textIsComplete && this.#audioIsComplete) {
-      const actualDuration = this.#totalAudioDuration;
-      if (actualDuration > 0 && this.#currentText.length > 0) {
-        this.#charsPerSecond = this.#currentText.length / actualDuration;
-      }
-    }
+  get playedText(): string {
+    return this.#playedText;
   }
 
-  #computeSleepInterval(): number {
-    return Math.min(Math.max(1 / this.#charsPerSecond, 0.0625), 0.5);
+  async close(interrupt: boolean) {
+    if (this.#closed) {
+      return;
+    }
+    this.#closed = true;
+    this.#interrupted = interrupt;
```

Comment:
Do we need to store the interrupted state inside the TextSynchronizer? This seems like it isn't its responsability vs the VPA

Reply:
this is used in the sentence synchronizer, because the sync is sometimes a bit too slow, and the text doesn't show up fully. the functionality now --
- if interrupted, it will stop the text
- if not interrupted, it will concatenate the rest of the unwritten text to the end